tests/internal/aws_compress.c

```diff
@@ -0,0 +1,490 @@
+                            &out_data_b64_len);
+            flb_free(out_data);
+            out_data = NULL;
+            if (!TEST_CHECK(ret == 0)) {
+                TEST_MSG("| Decompression failure");
+                out_data_b64 = flb_malloc(1); /* placeholder malloc */
+            }
+        }
+
+        ret = memcmp(tcase->expect_out_data_b64, out_data_b64, out_data_b64_len);
+        TEST_CHECK(ret == 0);
+        TEST_MSG("| Expected output(%s): %s",
+                (test_type == FLB_AWS_COMPRESS_TEST_TYPE_COMPRESS)
+                ? "b64" : "decompressed", tcase->expect_out_data_b64);
+        TEST_MSG("| Produced output(%s): %s",
+                (test_type == FLB_AWS_COMPRESS_TEST_TYPE_COMPRESS)
+                ? "b64" : "decompressed", out_data_b64);
+
+        len = strlen(tcase->expect_out_data_b64);
+        TEST_CHECK(len == out_data_b64_len);
+        TEST_MSG("| Expected length: %zu", len);
+        TEST_MSG("| Produced length: %zu", out_data_b64_len);
+
+        TEST_MSG("*- For input data: %s", tcase->in_data);
+
+        flb_free(out_data_b64);
+        ++tcase;
+    }
+}
+
+/* B64 check script copied from Monkey Auth Plugin */
+/* Copied from monkey/plugins/auth/base64.c */
+
+#include <monkey/mk_api.h>
+#if defined(MALLOC_JEMALLOC)
+#define __mem_alloc    mk_api->mem_alloc
+#define __mem_free     mk_api->mem_free
+#else
+#define __mem_alloc    malloc
+#define __mem_free     free
+#endif
+
+static const unsigned char base64_table[65] =
+	"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
+
+/**
+ * base64_encode - Base64 encode
+ * @src: Data to be encoded
+ * @len: Length of the data to be encoded
+ * @out_len: Pointer to output length variable, or %NULL if not used
+ * Returns: Allocated buffer of out_len bytes of encoded data,
+ * or %NULL on failure
+ *
+ * Caller is responsible for freeing the returned buffer. Returned buffer is
+ * nul terminated to make it easier to use as a C string. The nul terminator is
+ * not included in out_len.
+ */
+static unsigned char * base64_encode(const unsigned char *src, size_t len,
+			      size_t *out_len)
+{
+	unsigned char *out, *pos;
+	const unsigned char *end, *in;
+	size_t olen;
+	int line_len;
+
+	olen = len * 4 / 3 + 4; /* 3-byte blocks to 4-byte */
+	olen += olen / 72; /* line feeds */
+	olen++; /* nul termination */
+	if (olen < len)
+		return NULL; /* integer overflow */
+	if (mk_api != NULL) {
+		out = __mem_alloc(olen);
+	}
+	else {
+                out = __mem_alloc(olen);
+        }
+
+	if (out == NULL)
+		return NULL;
+
+	end = src + len;
+	in = src;
+	pos = out;
+	line_len = 0;
+	while (end - in >= 3) {
+		*pos++ = base64_table[in[0] >> 2];
+		*pos++ = base64_table[((in[0] & 0x03) << 4) | (in[1] >> 4)];
+		*pos++ = base64_table[((in[1] & 0x0f) << 2) | (in[2] >> 6)];
+		*pos++ = base64_table[in[2] & 0x3f];
+		in += 3;
+		line_len += 4;
+		if (line_len >= 72) {
+			*pos++ = '\n';
+			line_len = 0;
```

Comment:
This is a little different than our mbedtls implementation since it adds newlines. Our tests have short outputs where this newline is not utilized.

Reply:
Should we remove this or add a comment for the benefit of future maintainers?

Reply:
Yep, that sounds good.

Reply:
I removed the newline.